### DISCUSSION TITLE
How to auto-create a page-bundle with a preview image?

### DISCUSSION THREAD

**Discussion by rshmhrj** [estruyf/vscode-front-matter#461] [Q&A] [answered]
Hi @estruyf,
The [Congo theme](https://github.com/jpanther/congo) I'm using for Hugo just released a [new version](https://github.com/jpanther/congo/releases/tag/v2.4.0) with support for [thumbnails](https://jpanther.github.io/congo/samples/), [cover images](https://jpanther.github.io/congo/samples/rich-content/), and [feature images](https://jpanther.github.io/congo/docs/getting-started/#feature-cover-and-thumbnail-images).  It seems a page-bundle is required for it to work, not the `preview` frontmatter variable.  If I convert all current pages to page-bundles with the preview images inside the bundle, the theme should work.  I could then update the `preview` to point to the thumbnail inside the page-bundle to keep FrontMatter working.

How could I automate this process going forward, for new content?

I guess I could update the `.frontmatter/templates/default.md` file to auto-add the `preview` frontmatter.

The only manual step would be to take each of the new files that get created and convert it to a page bundle with either a default image or a custom image for that piece of content.  Do you think that could be automated with a [custom script](https://frontmatter.codes/docs/custom-actions#creating-a-content-script) which could get triggered whenever a new content is created?  I haven't looked at that option much, but just wondering if it you think it is possible and worth more exploration.

Also, FYI, here is the [Congo discussion](https://github.com/jpanther/congo/discussions/363#discussioncomment-4116307) with @jpanther where we are talking about the new feature.

Thanks for all your guidance and support!

**Comment by rshmhrj** [comment, marked as answer]
Going through the documentation, I found [content types](https://frontmatter.codes/docs/content-creation/content-types#content-type-properties), [fields](https://frontmatter.codes/docs/content-creation/fields#image), and [placeholders](https://frontmatter.codes/docs/content-creation/placeholders#custom-placeholders) which allows me to do something like this:
```json
  "frontMatter.taxonomy.contentTypes": [
    {
      "name": "default",
      "pageBundle": false,
      "previewPath": null,
      "template": "[[workspace]]/.frontmatter/templates/default.md",
      "fields": [ ... ]
    },
    {
      "name": "bundle with thumbnail",
      "pageBundle": true,
      "previewPath": null,
      "template": "[[workspace]]/.frontmatter/templates/default.md",
      "fields": [
        ...
        {
          "title": "Content preview",
          "name": "preview",
          "type": "image",
          "isPreviewImage": true,
          "default": "thumbnail.png"
        },
        ...
      ]
    },
    {
      "name": "bundle with feature image",
      "pageBundle": true,
      "previewPath": null,
      "template": "[[workspace]]/.frontmatter/templates/default.md",
      "fields": [
        ...
        {
          "title": "Content preview",
          "name": "preview",
          "type": "image",
          "isPreviewImage": true,
          "default": "feature.png"
        },
        ...
      ]
    }
  ],
```

Then I can choose the type when creating the content, and have it create as a pagebundle as well as auto populate the `preview` field in the frontmatter with either `feature` or `thumbnail`.  
![image](https://user-images.githubusercontent.com/750874/201509366-3615bab3-840c-49b6-a017-e81c11fb50c8.png)

![image](https://user-images.githubusercontent.com/750874/201509386-c6ed2450-62a5-4fd7-b43b-f0977b105940.png)

![image](https://user-images.githubusercontent.com/750874/201509400-c95d3b7e-8d09-4ee2-ac8f-b81e2f75ffb4.png)

![image](https://user-images.githubusercontent.com/750874/201509414-f6199ae9-4f84-48de-8697-8c8866ef61ec.png)

That gets me very far!  The only part that is missing is auto adding some default image, maybe as part of a [postScript](https://frontmatter.codes/docs/content-creation/content-types#content-type-properties) with a [custom bash script](https://frontmatter.codes/docs/custom-actions#bash-starter).

What I've tried so far:

```json
...
  "frontMatter.taxonomy.contentTypes": [
    ...
    {
      "name": "bundle with thumbnail",
      "pageBundle": true,
      "previewPath": null,
      "template": "[[workspace]]/.frontmatter/templates/default.md",
      "postScript": "copyLion-shell",
      "fields": [ ... ]
    },
    ...
  ],
  ...
  "frontMatter.custom.scripts": [
    {
      "id": "copyLion-shell",
      "title": "Copy Lion via shell",
      "script": "./scripts/copy_lion.sh",
      "command": "/bin/bash",
      "type": "content",
      "bulk": false,
      "output": "notification",
      "outputType": "text",
      "hidden": false
    },
  ]
```

contents of `./scripts/copy_lion.sh`
```bash
#!/bin/bash
echo "workspaceArg: $1"
echo "fileArg: $2"
echo "frontMatterArg: $3"

cp $1/assets/img/bw_lion.png $1/$2/thumbnail.png
```

But I'm getting some `undefined` error:
![image](https://user-images.githubusercontent.com/750874/201510363-a371d090-bf86-42bf-8470-e8da8ae71428.png)

@estruyf any thoughts on what might be causing the `undefined` error?

**Reply by estruyf** [reply]
@rshmhrj there was an issue with the error notification, it was reported in #462 - it should have been fixed. Can you try again to see what the actual error is it returns?

**Reply by estruyf** [reply]
Also, for `bash`, you should be able to just use `bash` as the `command` value instead of `/bin/bash`.

**Reply by rshmhrj** [reply]
That was definitely helpful!!  Arg `$2` included the filename, so needed to replace it:

`copylion.sh`
```bash
#!/bin/bash

echo "hi"
echo "workspaceArg: $1"
echo "fileArg: $2"
echo "frontMatterArg: $3"

outputFilePath=`echo $2 | sed 's#index.md#bw-lion_thumb.png#'`

cp $1/assets/thumb/bw-lion_thumb.png $outputFilePath
```

`frontmatter.json`
```json
  "frontMatter.custom.scripts": [
    {
      "id": "copyLion-shell",
      "title": "Copy Lion via shell",
      "script": "./scripts/copy_lion.sh",
      "command": "bash"
    }
  ]
```

```json
  "frontMatter.taxonomy.contentTypes": [
    ...
    {
      "name": "bundle with thumbnail",
      "pageBundle": true,
      "previewPath": null,
      "template": "[[workspace]]/.frontmatter/templates/default.md",
      "postScript": "copyLion-shell",
      "fields": [
        ...
        {
          "title": "Content preview",
          "name": "preview",
          "type": "image",
          "isPreviewImage": true,
          "default": "bw-lion_thumb.png"
        },
        ...
      ]
    }
  ],
```
